### PR TITLE
Refactor unary ops in ggml

### LINKED ISSUE #285: ggml : reduce the values of `ggml_op` enum
Here is another idea to consider for reducing the amount of unary ops and code duplication in `ggml.c`:

The following ops:

```c
    enum ggml_op {
        ...
        GGML_OP_ABS,
        GGML_OP_SGN,
        GGML_OP_NEG,
        GGML_OP_STEP,
        GGML_OP_TANH,
        GGML_OP_ELU,
        GGML_OP_RELU,
        GGML_OP_GELU,
        GGML_OP_GELU_QUICK,
        GGML_OP_SILU,
        ...
    };
```

Can all become:

```c
    enum ggml_unary_op {
        GGML_UNARY_OP_ABS,
        GGML_UNARY_OP_SGN,
        GGML_UNARY_OP_NEG,
        GGML_UNARY_OP_STEP,
        GGML_UNARY_OP_TANH,
        GGML_UNARY_OP_ELU,
        GGML_UNARY_OP_RELU,
        GGML_UNARY_OP_GELU,
        GGML_UNARY_OP_GELU_QUICK,
        GGML_UNARY_OP_SILU,    
    };

    GGML_API struct ggml_tensor * ggml_unary(
            struct ggml_context * ctx,
             struct ggml_tensor * a,
             enum ggml_unary_op   op);
```

---

### Original task (completed via #313)

Certain `ggml` operators can be combined into a singe operator with extra arguments.

For example:

- `GGML_OP_CONV_1D_S1_PH`
- `GGML_OP_CONV_1D_S2_PH`

can all become:

- `GGML_OP_CONV_1D`

with a signature of:

```c
    GGML_API struct ggml_tensor * ggml_conv_1d(
            struct ggml_context * ctx,
            struct ggml_tensor  * a,
            struct ggml_tensor  * b
                            int   stride,
                            int   padding);
```

Internally, we can have custom implementations of `ggml_compute_forward_conv_1d()` for specific values of the parameters and `GGML_ASSERT` for everything else that is not supported.

This will allow to extend the functionality of these operator more easily in the future

### PR DESCRIPTION
**Pull request by Ivan-Z**
Resolves #285. Removes duplicated code from unary op constructors to a shared ggml_unary. No changes to usage.

### CODE REVIEW

File: src/ggml.c [hunk 1]
```diff
@@ -14376,6 +14200,66 @@ static void ggml_compute_forward_win_unpart(
     }
 }
 
+//gmml_compute_forward_unary
```
Comment (ggerganov) [inline src/ggml.c:14203]:
```suggestion
// ggml_compute_forward_unary
```

File: src/ggml.c [hunk 2]
```diff
@@ -7538,6 +7322,47 @@ struct ggml_tensor * ggml_win_unpart(
...
+    bool is_node = false;
+
+    if (!inplace && (a->grad)) {
+        is_node = true;
+    }
+
+    struct ggml_tensor * result = inplace ? ggml_view_tensor(ctx, a) : ggml_dup_tensor(ctx, a);
+
+    struct ggml_tensor * b = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, 1);
+    ((int *) b->data)[0] = op;
```
Comment (ggerganov) [inline src/ggml.c:7341]: The parameter passing was just updated on `master`.
Try to utilize the new `op_params` array to pass this value

File: src/ggml.c [hunk 3]
```diff
@@ -14376,6 +14200,66 @@ static void ggml_compute_forward_win_unpart(
...
+            {
+                ggml_compute_forward_relu(params, src0, dst);
+            } break;
+        case GGML_UNARY_OP_GELU:
+            {
+               GGML_ASSERT(false); // TODO: not implemented
+            } break;
+        case GGML_UNARY_OP_GELU_QUICK:
+            {
+               GGML_ASSERT(false); // TODO: not implemented
```
Comment (ggerganov) [inline src/ggml.c:14250]: Why are `GELU` and `GELU_QUICK` not implemented?

File: src/ggml.c [hunk 4]
```diff
@@ -15967,6 +15748,80 @@ static void ggml_compute_backward(struct ggml_context * ctx, struct ggml_tensor
             } break;
         case GGML_OP_WIN_PART:
         case GGML_OP_WIN_UNPART:
+        case GGML_OP_UNARY:
+            {
+                switch((enum ggml_unary_op)tensor->src[0]) {
```
Comment (ggerganov) [inline src/ggml.c:15753]: This should be `tensor->src[1]` I think

File: src/ggml.c [hunk 5]
```diff
@@ -16459,21 +16314,34 @@ struct ggml_cplan ggml_graph_plan(struct ggml_cgraph * cgraph, int n_threads) {
...
                     n_tasks = 1;
                 } break;
-            case GGML_OP_MUL:
-            case GGML_OP_GELU:
-            case GGML_OP_GELU_QUICK:
-            case GGML_OP_SILU:
+
+            case GGML_OP_UNARY:
+                {
+                    switch ((enum ggml_unary_op)node->src[0]) {
```
Comment (ggerganov) [inline src/ggml.c:16323]: Should be `node->src[1]`

**Review by ggerganov — CHANGES_REQUESTED** [review]
Looks good, but a few things need fixes.

I'll push a change now to finalize this

**Review by ggerganov — APPROVED** [review]
What is missing now is to figure out a neat way to print and dump the names of the unary ops, because now it will simply print `unary` which is not very useful for debugging purposes.

@slaren Do you think this change is OK?

File: src/ggml.c [hunk 6]
```diff
@@ -3811,7 +3803,7 @@ static const char * GGML_OP_NAME[GGML_OP_COUNT] = {
     "CROSS_ENTROPY_LOSS_BACK",
 };
 
-static_assert(GGML_OP_COUNT == 68, "GGML_OP_COUNT != 68");
+static_assert(GGML_OP_COUNT == 59, "GGML_OP_COUNT != 68");
```
Comment (slaren) [inline src/ggml.c:3806]: The string needs to be updated as well.

File: src/ggml.c [hunk 7]
```diff
@@ -4958,6 +4961,14 @@ float * ggml_get_data_f32(const struct ggml_tensor * tensor) {
     return (float *)(tensor->data);
 }
 
+enum ggml_unary_op ggml_get_unary_op(const struct ggml_tensor * tensor) {
+    return (enum ggml_unary_op) ggml_get_op_params_i32(tensor, 0);
+}
+
+void ggml_set_unary_op(struct ggml_tensor * tensor, enum ggml_unary_op op) {
+    ggml_set_op_params_i32(tensor, 0, (int32_t) op);
+}
```
Comment (slaren) [inline src/ggml.c:4970]: A check such as `GGML_ASSERT(tensor->op == GGML_OP_UNARY)` would be good here.

Comment (slaren) [reply src/ggml.c:4970]: I am not sure if we really need `ggml_set_unary_op`, though.

**Comment by slaren** [comment]
A function that returns the name of a unary op would be useful. Maybe also one that returns either the name of the op or the name of the unary. Something like this:
```C
if (op == GGML_OP_UNARY) {
    return ggml_unary_op_name(t);
}
return ggml_op_name(t);
```

**Comment by Ivan-Z** [comment]
Thanks for the help @ggerganov @slaren. Do we want to make a similar change for binary ops? There are only a few right now, but might be worth making the change now if ggml intends on supporting more in the future.

**Comment by ggerganov** [comment]
Let's use the unary API for a while and decide later depending on if we like the API or not

**Pull request merged**